pkg/backend/httpstate/state.go

```diff
@@ -48,48 +48,51 @@ type tokenSource struct {
 	done     chan bool
 }
 
-func newTokenSource(ctx context.Context, token string, backend *cloudBackend, update client.UpdateIdentifier,
+func newTokenSource(ctx context.Context, initialToken string, backend *cloudBackend, update client.UpdateIdentifier,
 	duration time.Duration) (*tokenSource, error) {
 
-	// Perform an initial lease renewal.
-	newToken, err := backend.client.RenewUpdateLease(ctx, update, token, duration)
-	if err != nil {
-		return nil, err
-	}
-
 	requests, done := make(chan tokenRequest), make(chan bool)
 	go func() {
 		// We will renew the lease after 50% of the duration has elapsed to allow more time for retries.
 		ticker := time.NewTicker(duration / 2)
 		defer ticker.Stop()
 
+		token := initialToken
+		var currentError error
+
+		renew := func() {
+			newToken, err := backend.client.RenewUpdateLease(ctx, update, token, duration)
+
+			// If we get an error from the backend, leave `err` set and surface it during
+			// the next request for a lease token.
+			if err != nil {
+				logging.V(3).Infof("error renewing lease: %v", err)
+				currentError = fmt.Errorf("renewing lease: %w", err)
+
+				ticker.Stop()
+			} else {
+				token = newToken
+			}
+		}
```

Comment:
I wouldn't use a closure to set stack locals. I think that's a little departed from idiomatic Go. I'd prefer if the function returned the two values instead.
Then, we can move them closer to where they're used, inside the `for` loop.

Maybe a pattern like this:

```golang
for {
    var token, err = initialToken, nil
    select {
    case <-ticker.C:
        token, err = renew()
```

I'm not sure this is much better tbh, since `renew` still closes over `ctx`, `update`, `token`, `duration`, `ticker`, and `backend`. I'd prefer another top-level function for `renew` but that's a _lot_ of parameters, so maybe not!

Reply:
I'd like to learn more idiomatic Go but I worry about the suggestion:

```
for {
    var token, err = initialToken, nil
```

Would this not reset token to initialToken on *every* iteration? That's not what we want really - we want the state to progress `token1, token2, token3, etc`. 

I can tinker with moving this out to a helper struct.